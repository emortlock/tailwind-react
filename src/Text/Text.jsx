import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import { TailwindContext } from '../context'

const Text = ({
  children,
  className,
  is,
  size,
  p,
  lead,
  bold,
  italic,
  ...rest
}) => {
  const Component = is === 'span' && p ? 'p' : is

  return (
    <TailwindContext.Consumer>
      {theme => (
        <Component
          {...rest}
          className={classnames(
            'leading-normal',
            `font-${theme.text.family.body}`,
            `text-${
              theme.text.size.body[
                (lead ? theme.text.size.body.length : size) - 1
              ]
            }`,
            `text-${theme.textColors.body}`,
            p && `mb-${theme.spacing.md}`,
            bold && 'font-bold',
            italic && 'italic',
            className,
          )}
        >
          {children}
        </Component>
      )}
    </TailwindContext.Consumer>
  )
}

Text.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  size: PropTypes.number,
  p: PropTypes.bool,
  lead: PropTypes.bool,
  bold: PropTypes.bool,
  italic: PropTypes.bool,
}

Text.defaultProps = {
  children: undefined,
  className: undefined,
  is: 'span',
  size: 1,
  p: false,
  lead: false,
  bold: false,
  italic: false,
}

export default Text
