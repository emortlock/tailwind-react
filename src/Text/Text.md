Standard inline or body copy:

```jsx
<Text>Hello World</Text>
```

Lead text:

```jsx
<Text p lead>
  Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor
  incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis
  nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.
  Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore
  eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt
  in culpa qui officia deserunt mollit anim id est laborum.
</Text>
```

Font styles:

```jsx
<Text p>
  Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor
  incididunt ut labore et dolore magna aliqua.{' '}
  <Text bold>Ut enim ad minim veniam</Text>, quis nostrud exercitation ullamco
  laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in
  reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.{' '}
  <Text italic>Excepteur sint occaecat cupidatat non proident</Text>, sunt in
  culpa qui officia deserunt mollit anim id est laborum.
</Text>
```
