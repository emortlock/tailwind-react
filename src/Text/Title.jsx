import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import { TailwindContext } from '../context'

const Title = ({
  children,
  className,
  is,
  size,
  subtitle,
  flush,
  h,
  ...rest
}) => {
  const level = h || Math.max(7 - size, 1)
  const Component = is || `h${level}`
  let ariaProps = {}

  if (!subtitle && Component !== 'string' && !/h[1-6]/i.test(Component)) {
    ariaProps = {
      role: 'heading',
      'aria-level': level,
    }
  }
  return (
    <TailwindContext.Consumer>
      {theme => (
        <Component
          {...rest}
          {...ariaProps}
          className={classnames(
            `font-${theme.text.family[subtitle ? 'subtitle' : 'title']}`,
            'leading-tight',
            `text-${theme.text.size.title[size - 1]}`,
            !subtitle && `text-${theme.textColors.emphasis} font-bold`,
            subtitle && `text-${theme.textColors.body} font-medium`,
            !flush && `mb-${theme.spacing.md}`,
            className,
          )}
        >
          {children}
        </Component>
      )}
    </TailwindContext.Consumer>
  )
}

Title.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  size: PropTypes.number,
  subtitle: PropTypes.bool,
  flush: PropTypes.bool,
  h: PropTypes.number,
}

Title.defaultProps = {
  children: undefined,
  className: undefined,
  is: undefined,
  size: 4,
  subtitle: false,
  flush: false,
  h: undefined,
}

export default Title
