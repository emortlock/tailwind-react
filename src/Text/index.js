export { default as Subtitle } from './Subtitle'
export { default as Text } from './Text'
export { default as Title } from './Title'
