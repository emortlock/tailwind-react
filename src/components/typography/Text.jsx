import React from 'react'
import PropTypes from 'prop-types'

import { withTheme } from '../theme'
import { BaseComponent, propTypes } from '../tailwind'
import { getAsArray } from '../utils'

const Text = ({
  theme,
  children,
  is,
  size,
  lead,
  bold,
  italic,
  brand,
  paragraph,
  text,
  ...rest
}) => {
  const isParagraph = is === 'p' || paragraph
  return (
    <BaseComponent
      is={is}
      leading="normal"
      font={[theme.text.family.body, bold && 'bold']}
      text={[
        theme.text.size.body[(lead ? theme.text.size.body.length : size) - 1],
        !brand && theme.textColors.body,
        ...getAsArray(text),
      ]}
      m={isParagraph ? { b: theme.spacing.md } : undefined}
      italic={italic || undefined}
      {...rest}
    >
      {children}
    </BaseComponent>
  )
}

Text.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  children: PropTypes.node,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  size: PropTypes.number,
  lead: PropTypes.bool,
  bold: PropTypes.bool,
  italic: PropTypes.bool,
  brand: PropTypes.bool,
  paragraph: PropTypes.bool,
  text: propTypes.text,
}

Text.defaultProps = {
  children: undefined,
  is: 'span',
  size: 2,
  paragraph: false,
  lead: false,
  bold: false,
  italic: false,
  brand: false,
  text: undefined,
}

export { Text as component }
export default withTheme(Text)
