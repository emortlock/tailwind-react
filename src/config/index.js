export { default as defaultConfig } from './defaultConfig'
export { default as TailwindConfig } from './TailwindConfig'
export { default as withConfig } from './withConfig'
