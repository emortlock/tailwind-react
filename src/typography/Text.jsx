import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import { withTheme } from '../theme'
import {
  getTailwindClassNames,
  tailwindProps,
  tailwindPropTypes,
} from '../tailwind'
import { filterProps } from '../utils'

const Text = ({
  theme,
  children,
  className,
  is,
  size,
  lead,
  bold,
  italic,
  brand,
  paragraph,
  ...rest
}) => {
  const Component = is
  const userClassNames = classnames(getTailwindClassNames(rest), className)

  const isParagraph = is === 'p' || paragraph

  return (
    <Component
      {...filterProps(rest, tailwindProps)}
      className={classnames(
        'leading-normal',
        `font-${theme.text.family.body}`,
        `text-${
          theme.text.size.body[(lead ? theme.text.size.body.length : size) - 1]
        }`,
        !brand && `text-${theme.textColors.body}`,
        isParagraph && `mb-${theme.spacing.md}`,
        bold && 'font-bold',
        italic && 'italic',
        userClassNames,
      )}
    >
      {children}
    </Component>
  )
}

Text.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  children: PropTypes.node,
  className: PropTypes.string,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  size: PropTypes.number,
  lead: PropTypes.bool,
  bold: PropTypes.bool,
  italic: PropTypes.bool,
  brand: PropTypes.bool,
  paragraph: PropTypes.bool,
  ...tailwindPropTypes,
}

Text.defaultProps = {
  children: undefined,
  className: undefined,
  is: 'span',
  size: 2,
  paragraph: false,
  lead: false,
  bold: false,
  italic: false,
  brand: false,
}

export { Text as component }
export default withTheme(Text)
