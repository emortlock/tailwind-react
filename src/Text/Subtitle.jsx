import { compose, setDisplayName, withProps } from 'recompose'
import Title from './Title'

export default compose(
  setDisplayName('Subtitle'),
  withProps({ subtitle: true }),
)(Title)
