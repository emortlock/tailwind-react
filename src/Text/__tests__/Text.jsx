import React from 'react'
import { shallow } from 'enzyme'

import { component as Text } from '../Text'
import { defaultConfig } from '../../config'

const setup = (testProps = {}) => {
  const props = Object.assign({ config: defaultConfig }, testProps)

  const wrapper = shallow(<Text {...props}>Lorem ipsum dolor sit amet</Text>)

  return {
    props,
    wrapper,
  }
}

describe('Text', () => {
  it('renders matching snapshot', () => {
    const { wrapper } = setup({ p: true })

    expect(wrapper).toMatchSnapshot()
  })
})
