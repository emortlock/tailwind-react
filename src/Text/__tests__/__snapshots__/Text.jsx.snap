// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text renders matching snapshot 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
